### ISSUE TITLE
AST::clone() is broken for some AST node. This is a bug. 

### ISSUE THREAD

**Issue by qoega** [ClickHouse/ClickHouse#15283]
```
base before fuzz: SELECT * REPLACE(i + 1048575 AS i) APPLY(sum) FROM columns_transformers
base after fuzz: SELECT * REPLACE(i + 256 AS i) APPLY(sum) FROM columns_transformers
dump before fuzz:
SelectWithUnionQuery, 0x7f4618c3fc98
-ExpressionList, 0x7f4618d971e8
--SelectQuery, 0x7f4618c6c418
---ExpressionList, 0x7f4618d97828
----Asterisk, 0x7f4618b89858
-----ColumnsReplaceTransformer, 0x7f461a5c7ca8
------ColumnsReplaceTransformer::Replacement, 0x7f4618cfa9c8
-------Function_plus, 0x7f4618d5add8
--------ExpressionList, 0x7f4618c4d3d8
---------Identifier_i, 0x7f4618ce2f18
---------Literal_Int64_1024, 0x7f4618c5e720
-------Function_plus, 0x7f4618d5b058
--------ExpressionList, 0x7f4618d97918
---------Identifier_i, 0x7f4618ce2cd8
---------Literal_Int64_-9223372036854775808, 0x7f4618c5e520
-------Function_plus, 0x7f4618d5a298
--------ExpressionList, 0x7f4618d97a08
---------Identifier_i, 0x7f4618ce3518
---------Literal_Int64_1048575, 0x7f4618c5ea20
-----ColumnsApplyTransformer, 0x7f4618b89a38
---TablesInSelectQuery, 0x7f4618d97378
----TablesInSelectQueryElement, 0x7f4618d51998
-----TableExpression, 0x7f461a563c58
------Identifier_columns_transformers, 0x7f4618ce2918

dump of cloned ast:
SelectWithUnionQuery, 0x7f4618d5a3d8
-ExpressionList, 0x7f4618d970f8
--SelectQuery, 0x7f4618d51b18
---ExpressionList, 0x7f4618c2a668
----Asterisk, 0x7f4618b89678
-----ColumnsReplaceTransformer, 0x7f4618b79b88
------ColumnsReplaceTransformer::Replacement, 0x7f4618cfabf8
-------Function_plus, 0x7f4618d5add8
--------ExpressionList, 0x7f4618c4d3d8
---------Identifier_i, 0x7f4618ce2f18
---------Literal_Int64_1024, 0x7f4618c5e720
-------Function_plus, 0x7f4618d5b058
--------ExpressionList, 0x7f4618d97918
---------Identifier_i, 0x7f4618ce2cd8
---------Literal_Int64_-9223372036854775808, 0x7f4618c5e520
-------Function_plus, 0x7f4618d5a298
--------ExpressionList, 0x7f4618d97a08
---------Identifier_i, 0x7f4618ce3518
---------Literal_Int64_1048575, 0x7f4618c5ea20
-------Function_plus, 0x7f4618d5a0b8
--------ExpressionList, 0x7f4618d97418
---------Identifier_i, 0x7f4618cd89d8
---------Literal_Int64_1048575, 0x7f4618c5ee20
-----ColumnsApplyTransformer, 0x7f4618b898b8
---TablesInSelectQuery, 0x7f4618d97058
----TablesInSelectQueryElement, 0x7f4618d51718
-----TableExpression, 0x7f4618d5ac98
------Identifier_columns_transformers, 0x7f4618ce3458

dump after fuzz:
SelectWithUnionQuery, 0x7f4618c3fc98
-ExpressionList, 0x7f4618d971e8
--SelectQuery, 0x7f4618c6c418
---ExpressionList, 0x7f4618d97828
----Asterisk, 0x7f4618b89858
-----ColumnsReplaceTransformer, 0x7f461a5c7ca8
------ColumnsReplaceTransformer::Replacement, 0x7f4618cfa9c8
-------Function_plus, 0x7f4618d5add8
--------ExpressionList, 0x7f4618c4d3d8
---------Identifier_i, 0x7f4618ce2f18
---------Literal_Int64_1024, 0x7f4618c5e720
-------Function_plus, 0x7f4618d5b058
--------ExpressionList, 0x7f4618d97918
---------Identifier_i, 0x7f4618ce2cd8
---------Literal_Int64_-9223372036854775808, 0x7f4618c5e520
-------Function_plus, 0x7f4618d5a298
--------ExpressionList, 0x7f4618d97a08
---------Identifier_i, 0x7f4618ce3518
---------Literal_Int64_256, 0x7f4618c5ea20
-----ColumnsApplyTransformer, 0x7f4618b89a38
---TablesInSelectQuery, 0x7f4618d97378
----TablesInSelectQueryElement, 0x7f4618d51998
-----TableExpression, 0x7f461a563c58
------Identifier_columns_transformers, 0x7f4618ce2918
IAST::clone() is broken for some AST node. This is a bug. The original AST ('dump before fuzz') and its cloned copy ('dump of cloned AST') refer to the same nodes, which must never happen. This means that their parent node doesn't implement clone() correctly.clickhouse-client: ../programs/client/Client.cpp:1173: const char *DB::Client::processWithFuzzing(const DB::String &): Assertion `false' failed.
```

**Comment by akuzm** [comment]
https://github.com/ClickHouse/ClickHouse/issues/14695

**Comment by amosbird** [comment]
Heh, I've mis-cloned AST twice...

**Comment by akuzm** [comment]
> Heh, I've mis-cloned AST twice...

You think it's something different? I thought it was the same and wanted to close the task, but apparently forgot to press the button.

**Comment by akuzm** [comment]
I tried with today's master and it still fails, so probably different. It's simple to check locally if you run `clickhouse-client --query-fuzzer-runs=1000` and enter the failing query.

```
base before fuzz: SELECT * REPLACE(i + 1048575 AS i) APPLY(sum) FROM columns_transformers
base after fuzz: SELECT * REPLACE(i + 1024 AS i) APPLY(sum) FROM columns_transformers
dump before fuzz:
SelectWithUnionQuery, 0x7f573187b858
-ExpressionList, 0x7f5732a22bb8
--SelectQuery, 0x7f57317e6e98
---ExpressionList, 0x7f5732a22c08
----Asterisk, 0x7f5731a50678
-----ColumnsReplaceTransformer, 0x7f5732a22a78
------ColumnsReplaceTransformer::Replacement, 0x7f5731869578
-------Function_plus, 0x7f573187b2b8
--------ExpressionList, 0x7f5732a22ac8
---------Identifier_i, 0x7f5731972918
---------Literal_UInt64_1048575, 0x7f573180a820
-----ColumnsApplyTransformer, 0x7f5731a505b8
---TablesInSelectQuery, 0x7f5732a22938
----TablesInSelectQueryElement, 0x7f57317e6f18
-----TableExpression, 0x7f573187b358
------Identifier_columns_transformers, 0x7f57319726d8

dump of cloned ast:
SelectWithUnionQuery, 0x7f5732b349d8
-ExpressionList, 0x7f5732a227f8
--SelectQuery, 0x7f5731919598
---ExpressionList, 0x7f5732a22848
----Asterisk, 0x7f573199d2b8
-----ColumnsReplaceTransformer, 0x7f5732a22708
------ColumnsReplaceTransformer::Replacement, 0x7f5731869188
-------Function_plus, 0x7f573187b2b8
--------ExpressionList, 0x7f5732a22ac8
---------Identifier_i, 0x7f5731972918
---------Literal_UInt64_1048575, 0x7f573180a820
-------Function_plus, 0x7f573187a458
--------ExpressionList, 0x7f5732a22578
---------Identifier_i, 0x7f57319720d8
---------Literal_UInt64_1048575, 0x7f573180a420
-----ColumnsApplyTransformer, 0x7f573199d6d8
---TablesInSelectQuery, 0x7f57322d69b8
----TablesInSelectQueryElement, 0x7f57317e6a98
-----TableExpression, 0x7f573187a958
------Identifier_columns_transformers, 0x7f57318f2dd8

dump after fuzz:
SelectWithUnionQuery, 0x7f573187b858
-ExpressionList, 0x7f5732a22bb8
--SelectQuery, 0x7f57317e6e98
---ExpressionList, 0x7f5732a22c08
----Asterisk, 0x7f5731a50678
-----ColumnsReplaceTransformer, 0x7f5732a22a78
------ColumnsReplaceTransformer::Replacement, 0x7f5731869578
-------Function_plus, 0x7f573187b2b8
--------ExpressionList, 0x7f5732a22ac8
---------Identifier_i, 0x7f5731972918
---------Literal_Int64_1024, 0x7f573180a820
-----ColumnsApplyTransformer, 0x7f5731a505b8
---TablesInSelectQuery, 0x7f5732a22938
----TablesInSelectQueryElement, 0x7f57317e6f18
-----TableExpression, 0x7f573187b358
------Identifier_columns_transformers, 0x7f57319726d8
IAST::clone() is broken for some AST node. This is a bug. The original AST ('dump before fuzz') and its cloned copy ('dump of cloned AST') refer to the same nodes, which must never happen. This means that their parent node doesn't implement clone() correctly.clickhouse-client: /home/akuzm/ch1/ch/programs/client/Client.cpp:1173: const char *DB::Client::processWithFuzzing(const DB::String &): Assertion `false' failed.
```

**Comment by tavplubix** [comment]
Fixed in #15302
